File: server/src/utils/geoUtils.js
export const calculateDistance = (
  lat1,
  lon1,
  lat2,
  lon2
) => {
  const R = 6371;

  const dLat =
    ((lat2 - lat1) * Math.PI) / 180;

  const dLon =
    ((lon2 - lon1) * Math.PI) / 180;

  const a =
    Math.sin(dLat / 2) *
      Math.sin(dLat / 2) +
    Math.cos(
      (lat1 * Math.PI) / 180
    ) *
      Math.cos(
        (lat2 * Math.PI) / 180
      ) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);

  const c =
    2 * Math.atan2(
      Math.sqrt(a),
      Math.sqrt(1 - a)
    );

  return R * c;
};

export const generateBoundingBox = (
  lat,
  lng,
  radiusKm = 10
) => {
  const latChange =
    radiusKm / 111;

  const lngChange =
    radiusKm /
    (111 *
      Math.cos((lat * Math.PI) / 180));

  return {
    north: lat + latChange,
    south: lat - latChange,
    east: lng + lngChange,
    west: lng - lngChange
  };
};

File: ml-service/app/utils/geo_utils.py
from math import radians
from math import sin
from math import cos
from math import sqrt
from math import atan2

from app.core.logger import logger

def calculate_distance(
    lat1,
    lon1,
    lat2,
    lon2
):
    try:
        logger.info(
            "Calculating geographic distance..."
        )

        R = 6371

        dlat = radians(
            lat2 - lat1
        )

        dlon = radians(
            lon2 - lon1
        )

        a = (
            sin(dlat / 2) ** 2
            +
            cos(radians(lat1))
            *
            cos(radians(lat2))
            *
            sin(dlon / 2) ** 2
        )

        c = 2 * atan2(
            sqrt(a),
            sqrt(1 - a)
        )

        return R * c

    except Exception as e:
        logger.error(str(e))

        raise Exception(
            "Geo calculation failed"
        )